### DISCUSSION TITLE
Feature roadmap

### DISCUSSION THREAD

**Discussion by daniel-brenot** [Leaflet/Leaflet#7729] [Ideas]
What features do we want to see after support has been dropped for Internet Explorer in [v2.0](https://github.com/Leaflet/Leaflet/milestone/39)?

Please use one comment per feature request if you have multiple feature requests.

**Comment by Falke-Design** [comment]
**Validation of methods arguments**

**Reply by mourner** [reply]
Personally I think this adds a lot of verbose code for very little benefit. Popular JS libraries usually don't strictly validate arguments except for very common mistakes, otherwise they would be much bigger.

**Reply by mathe42** [reply]
When we had "offical" types files that would be 100% not needed.

**Reply by Falke-Design** [reply]
Look into #7128 or a current usecase would be #7432. If a string is used instead of interges the L.Point() calculation goes wrong. But how to check correctly if it is a number? See https://github.com/Leaflet/Leaflet/pull/7432#discussion_r565217858

> adds a lot of verbose code

Maybe we should define Validation-Util functions and add them only at certain places.

> When we had "offical" types files that would be 100% not needed.

What are "offical" types? We have them already in the documentation but throws no error if the wrong type is passed. Same if TypeScript is not used.

**Reply by mathe42** [reply]
Typedefs in comment are removed at build. With .d.ts files in npm package (= "offical") users would get autocomplete in there ide. Even js users!

**Reply by mathe42** [reply]
Some idea: have a dev and a production build. In dev checks are done in prod checks are skiped (and no extra code!)

**Reply by mourner** [reply]
For simple cases where we get a string instead of a number, we can usually just do +arg. I would definitely not want to clutter the code with a ton of validation utility calls. I agree that a nice strict type coverage can somewhat fulfill the same need in TS codebases.

**Comment by Falke-Design** [comment]
**TypeScript .d.ts file**

Very useful would be to refactor Events because there is no easy handling in TypeScript definition. See: https://github.com/geoman-io/leaflet-geoman/issues/893
What we have done in Leaflet-Geoman: [leaflet-geoman.d.ts](https://github.com/geoman-io/leaflet-geoman/blob/master/leaflet-geoman.d.ts)

**Reply by mathe42** [reply]
Events can be done with a Eventmapinterface. And Generics. Currently looking for an example

**Reply by mathe42** [reply]
examples:
1. https://stackoverflow.com/questions/65668969/event-maps-and-type-guards
2. https://github.com/microsoft/TypeScript/blob/main/lib/lib.dom.d.ts#L6401

**Reply by mathe42** [reply]
If you are interessted in adding typings in this repo. I would migrate the existing types. And I would try to keep them up to date when pr are merged.

**Reply by mathe42** [reply]
I would split the dts files so we have for (some) src files typings. That would imporove ide support while developing leaflet.

**Reply by mathe42** [reply]
see #7733

**Reply by jonkoops** [reply]
I am a huge proponent of this feature. It would greatly improve the out of the box experience of TypeScript users and those who simply expect some level of auto-completion in their IDE.

**Reply by ryan-morris** [reply]
@mathe42 just to clarify, are you saying that events can be done with the EventMapInterface now alongside the existing `@types/leaflet`, or are you recommending shifting them to that style declaration?

**Reply by mathe42** [reply]
I think the types from `@types/leaflet` should be added to this reposetory. The talk about the EventMap is only that @Falke-Design thought that with a change to the Evented class we can get better typings (and much smaller) but this is possible with the "EventMap-Interface"-Syntax. 

I'm working on it in #7733.

This would mean that @types/leaflet will become deprecated. But keeping it near the source code means less errors and maintaining them is easier.

**Comment by Falke-Design** [comment]
**Breaking compatibility with IE except IE11 & VML deprecation**

**Reply by mourner** [reply]
Definitely the main thing for v2. Break compatibility with old IE so that we could make maintaining Leaflet easier and modernize the codebase further.

**Reply by mathe42** [reply]
Also a general deprecation plan would be nice. ( "we support browsers that are not more than X years old" - or something like that.

**Reply by jonkoops** [reply]
Why not go all the way and drop support for Internet Explorer entirely? There are a lot of useful APIs (such as [Promises](https://caniuse.com/promises), [URLSearchParams](https://caniuse.com/urlsearchparams), etc) that are available only in modern browsers.

**Reply by Eschon** [reply]
I'd love to just drop support for IE11 everywhere, but I still have to support it in most projects I'm working on. And there are probably lots of devs in similar situations. Dropping older Versions should already be a huge improvement but I think it makes sense to support IE11 just a bit longer.

**Reply by mathe42** [reply]
I agree with @Eschon but I think it would be nice to set a date when to drop support. But even with IE support we get a lot of improfments in the codebase.

**Reply by Eschon** [reply]
According to [Microsoft](https://docs.microsoft.com/en-us/lifecycle/products/internet-explorer-11) IE11 Support will mostly end in June 2022. Dropping support in leaflet could be planned around that date.

**Reply by jonkoops** [reply]
I guess we could do an approach of 'bring your own polyfill', where if you are using an ancient browser like IE it's your responsibility to ship the correct polyfills. Personally I'd drop it due to the lack of support for ES2015 syntax, which is where I would put the cut-off point.

Those users who care about supporting Internet Explorer (an increasing minority) could also keep using the 1.x version without much implications.

**Reply by Falke-Design** [reply]
Maybe a "What polyfills do I need" would be useful. Is there a way to check which polyfills are needed and add them into a document?

**Reply by Eschon** [reply]
Instructing devs to include certain polyfills for IE11 would be fine for me. I just think that there should be an officially supported way to run leaflet in IE11. At least for now.

**Comment by daniel-brenot** [comment]
Writing the library using newer ES6 syntax.

**Reply by daniel-brenot** [reply]
This should still let us support older browsers using rollup to create bundles that work for older or newer versions

**Reply by mathe42** [reply]
I could not get the current class syntaxt (with no API change) in es6 that can be transpiled to es3 as I needed `classinstance.__proto__` or `classinstance.constructor.prototype`

**Reply by mathe42** [reply]
Note when dperecating old IE ignore my last comment

**Reply by jonkoops** [reply]
I think that 'using new ES6 syntax' is too broad, instead we should focus on what features will provide benefits to the community. For example aligning the custom class implementation with standardized classes.

**Reply by mathe42** [reply]
Maybe a better statement would be **allow es6 syntax (and above) in PR** and **update code where new syntax would make the code more readable**.

**Comment by Falke-Design** [comment]
**Polyfill deprecation. Half of the methods in L.Util would go away.**

I personal use a lot of L.Util functions because they are clear and easy to read. Maybe we don't need to remove L.Util, we can remove only the polyfills.

**Reply by daniel-brenot** [reply]
Polyfills can always be provided through libraries that offer it like babel if needed.

**Reply by mathe42** [reply]
No babel dosen't polyfill back to IE (es3) only to es5.

**Reply by daniel-brenot** [reply]
I've done this before. I'm currently supporting IE11 on a corporate site

**Reply by mathe42** [reply]
Yes but IE11 = es5. es3 = IE 7-10

**Reply by mathe42** [reply]
So with babel you can only suport IE 11 but not older versions

**Reply by daniel-brenot** [reply]
Aah i see. I didn't think you were talking about 7-10, as it seems everyone is happy not supporting those versions.

**Comment by Falke-Design** [comment]
**Getting rid of global switches (L_NO_TOUCH and L_DISABLE_3D)**

**Reply by mathe42** [reply]
I would say if we could get more of this switches we could make IE support optional as not needed code can be removed for a "No IE Build" by bundlers at build time of users.

**Reply by daniel-brenot** [reply]
Or produce more bundles. One for older browsers and one for newer. This practice is quite common, and easier to implement since we wouldn't need to write our own polyfills

**Reply by mathe42** [reply]
Your options requires same code changes as my, just time when code is removed is changed

**Reply by IvanSanchez** [reply]
Instead of switches, it's currently possible to write values to properties of `L.Browser`. Feels cleaner to keep allowing this rather than to keep global switches.

**Reply by mathe42** [reply]
But that would not eliminate dead code.

**Reply by IvanSanchez** [reply]
On the contrary: `L_NO_TOUCH` can go away because people can set `L.Browser.touch = false`. Idem with `L_DISABLE_3D` and `L.Browser.any3d`.

**Comment by Falke-Design** [comment]
**Removing CSS 3D acceleration hacks that likely haven't been relevant for years.**

**Comment by mathe42** [comment]
**Treeshaking**  in one of 2 ways:
1. With a lot of API breaking all nested exports e.g. (L.Controls....) has to get "top level exports" (this can be done in a "friendly" way where Plugins only habe to run a codemod to change Im/Exports)
2. With more switches for example L_NO_POLYGON that can be set to true at build time of the users of this lib

## Why?
If you look at 
![image](https://user-images.githubusercontent.com/24830662/139926857-37f16ba5-9026-4671-888e-00321196cb9e.png)
Leaflet wants to be modular (what it is not currently) and small (for what it does is 40K not much but it is a big lib!) and treesaking would help with both.

**Reply by mathe42** [reply]
Just thought about it 1. can be done without breaking change if we keep the current export but add for evrything a "top level export" no codemod required.

For users who don't use a bundler the filesize would increase a little.

**Comment by Falke-Design** [comment]
**Moving mixins added through L.<Class>.include(...)**
Eliminating the use of side effects code across classes — e.g. moving mixins added through L.<Class>.include(...) in favor of having them inline

**Reply by mathe42** [reply]
1. This will also remove all (or near all) Sideeffekt code from modules (So code that runs and does something outside the file at import of that file). That is good.
2. Should `include` be deprecated and removed in a later version?

**Reply by Falke-Design** [reply]
I really like the behaviour and functions of `extend` and `include`, it would be nice for plugins if they are not removed

**Reply by mathe42** [reply]
Was just wondering.

**Comment by Falke-Design** [comment]
**https links**

We should support all pages over https incl. the cdn #7667: https://cdn.leafletjs.com/leaflet/v1.7.1/leaflet.zip

**Reply by mourner** [reply]
This one I'm fixing right now.

**Comment by Falke-Design** [comment]
**Make more small releases**
Is a good sign for users/public (it show that the project is alive) and also liberate maintainers from the "stress" of making huge releases that may break too many things. And if a minor release break small things, another one can fix it shortly.

**Reply by mourner** [reply]
We should definitely do this, and I'm now working on radically simplifying the release process so that it could be effortless (a matter of just pushing a tag).

**Comment by Falke-Design** [comment]
**Stale Bot**

**Reply by mourner** [reply]
This one is not dependent on a major version, we can add it at any time.

**Comment by daniel-brenot** [comment]
Making leaflet not expose "L" to the global scope unless imported as a script tag in the webpage.
https://www.w3.org/wiki/JavaScript_best_practices#:~:text=a%20good%20thing.-,Avoid%20globals,are%20an%20incredibly%20bad%20idea.&text=If%20you%20have%20global%20variables,will%20overwrite%20your%20variables%2Ffunctions.

**Comment by Falke-Design** [comment]
**Changing the line endings CRLF / LF to the same for all files.**
#7409
This is will break some PR because they have some files with the wrong line break. **BUT** it would be possible to change all existing PRs because the maintainers has mostly the permission to edit the PRs.

So my suggestion (after we reduced the PRs a lot):
1. Freez PRs
2. make a branch with where all files have the same line break (or directly in the master)
3. update each PR from a maintainer and check if it works when comparing against the new branch. On a weekend I can do this and change each PR but I would need write access for this.
4. merge the branch into the master

**Reply by Falke-Design** [reply]
And converting smart-tab to 4-spaces: #3870

**Reply by simon04** [reply]
Here's a draft PR: https://github.com/Leaflet/Leaflet/pull/9664. Let's see how easy it is to rebase onto upcoming changes.

Regarding "but somewhere we discussed that it can destroy the git history" from https://github.com/Leaflet/Leaflet/pull/9652#discussion_r2051789597:

The history is kept, and git will create new objects in `.git/objects`. It is a matter of configuring the diffing/blaming, e.g. https://github.com/Leaflet/Leaflet/pull/9664/files?w=1 `w=1` hides white space changes. And `git blame` may be configured to ignore certain commit hashes, see https://git-scm.com/docs/git-blame#Documentation/git-blame.txt---ignore-revs-fileltfilegt

Going one step further, we could discuss to adopt an automated code formatter such as the popular prettier: https://prettier.io/

**Comment by Falke-Design** [comment]


**Reply by mourner** [reply]
Why? Marker shadows look nice.

**Comment by mathe42** [comment]
**esm output + dont export to window.L**

Not shure if this will/should happen but we should discusse this

**Comment by daniel-brenot** [comment]
ESLint to ensure project follows constant code style.
This can also ensure that errors with types are caught if we choose to integrate that, which would allow for using js to write the code, but still get validation that errors caused by an undefined or null pointer don't happen nearly as often.

Edit: Move eslint configuration to separate file from package.json. This doesn't need to be there

**Reply by IvanSanchez** [reply]
https://github.com/Leaflet/Leaflet/blob/bb04e9567d7afc7c9a503e7c8b21f0ea4196fca3/package.json#L43

**Comment by mourner** [comment]
One from @IvanSanchez which is worth considering: consistent handling of events across mouse and touch devices through `PointerEvent` API. Although it might be too big of a quest for v2 and could be addressed later (v3?). Also possibly something of interest for @johnd0e.

> PointerEvents as first-class citizens. This breaks compat with IE10, Chrome 54, FFX 58, Safari 12. Rewrite of the whole event logic, full removal of the touch event detection and polyfills.

**Reply by IvanSanchez** [reply]
There's some work done in the [`pointerify` branch](https://github.com/Leaflet/Leaflet/tree/pointerify). If we're willing to break backwards compat with Leaflet 1 code, this becomes deleting a lot of code.

**Comment by IvanSanchez** [comment]
**Extendable full-map-container DOM elements**:

Split some logic from L.Renderer into a "full-map-container DOM element". Specifically I'm talking about https://gitlab.com/IvanSanchez/Leaflet.GLMarkers/-/blob/048ba429/src/GLMarkers.js#L90-L126 . This would ease integration of whatever D3 or WebGL overlays devs desire.

The idea is that there should be **four** kinds of `L.Layer`:
* DOM elements anchored to a `LatLng` (i.e. markers)
* DOM elements anchored to a `LatLngBounds` (i.e. imageoverlay)
* Grids of DOM elements (`L.GridLayer`s)
* DOM elements anchored to the viewport (i.e. both the `L.Canvas` and `L.SVG` renderers)

I think this is not a lot of work - just splitting the right parts of `L.Renderer` into a abstract parent class.

This might pair nicely with getting rid of the CSS animations. The code for this full-map-container overlay would take care of calling `reset()` and `update()` to redraw data and reposition itself as deemed fit, so this offers the possibility of throttling redraws during non-CSS zoom animations.

**Comment by mathe42** [comment]
**Breaking Changes**

There are multiple ways how we can help users to update leaflet when breaking changes happen:
1. Add deprecation messages in a earlier version  to change to a new syntax or what the change is.
2. Add codemods that plugin-developers and leaflet users can run that detect where the change is breaking something and logs that / or fixes that automaticly.

**Comment by mathe42** [comment]
**Rewrite Plugin Guide with ESM in mind**
https://github.com/Leaflet/Leaflet/blob/master/PLUGIN-GUIDE.md

I'm not a fan of the "Put all things on the global `L`"-Concept. But I know that we can't expect this to change in the near future (=next years). But I think we should explain new plugins authors how they can have both put the things on the global L and a esm build that is better with Treeshaking and bundeling.

This can be done with a single simple rollup-Plugin.

**Reply by mathe42** [reply]
If you are interessted I would create a example Plugin that generates both bundles.

**Reply by daniel-brenot** [reply]
Why can't we expect this to change in the near future? I'm more of a fan of doing an API rewrite correctly the first time rather than having the users rewrite their API only to have to rewrite it AGAIN as soon as V3 comes out. If we know that V3 is going to be significantly different, it's better to make those significant changes now so that the V3 port is less of a shock.

**Reply by mathe42** [reply]
Because near all plugins need `window.L` so before we remove `window.L` a huge amount of Plugins have to change. This might be easy for the more used and actively maintained ones.

Also putting Leaflet on L is not a bad step. The bad step is that plugins do that too. And don't create there own name like `window.L_Plugin_XYZ`.


Also you expect a full rewrite to happen - I don't think so. You would lose the hole Plugin-Ecosystem.

**Reply by Falke-Design** [reply]
Why do you don't like that? For me personally this is a nice feature

**Reply by mathe42** [reply]
This prevents treeshaking and useage with build tools get complicated.

In one case I decided that I add the plugin into my code and modify it so it can work properly with my build setup

**Reply by daniel-brenot** [reply]
Plugins can still access the global L in cases where it is imported via the browser. But the way plugins work could instead be by initializing it by passing in the value for L.

**Reply by mathe42** [reply]
I would say that it is better to use es import syntax and generate 2 bundles one for bundelers where leaflet is external and one where it get leaflet from window.L

**Reply by Falke-Design** [reply]
Like 
```
import Geoman from 'leaflet-geoman'
L.Geoman = Geoman
```
This ok for me and we (Geoman) has this already on the todo, bevause many peoples has problem with auto initializing

**Reply by mathe42** [reply]
I think in source code plugins should:
1. Use import statement to get leaflet
2. Use export syntax

Than we have 2 output files
1. "Bundler" build with import and export statements keept the same.
2. a "Browser" build that get Leaflet from `window.L` and add exports to `window.L.PLUGIN_NAME` if `window.L` don't exist it throws a Error.

My problem with "just put in on L" is if it is not optional. And than I get problems when using for example SSR.
My proposal would also allow treeshaking when using the "Bundler" build.

[EDIT] If Plugins want to change more than `L.PLUGIN_NAME` this can be done with 2 entry points. Or a build time var like:

```js
export const XYZ = a;

if(__BROWSER_BUILD__) {
 window.L.Controls.XYZ = a; 
}
```

**Reply by jeffreykog** [reply]
Adding to this, it would be really awesome if `window` is not accessed on import of the module at all, so that Leaflet can also be used in a non-hacky way in situations where server-side rendering is used, like with Angular ngUniversal.

**Reply by daniel-brenot** [reply]
> Adding to this, it would be really awesome if `window` is not accessed on import of the module at all, so that Leaflet can also be used in a non-hacky way in situations where server-side rendering is used, like with Angular ngUniversal.

This is exactly our use case! This is part of why i want it to not be bound to the window object.

**Comment by Falke-Design** [comment]
We should remove the deprecated functions. There are a few warnings while running tests:

- `_flat`: Deprecated use of _flat, please use L.LineUtil.isFlat instead.
- `L.Mixin.Events`: Deprecated include of L.Mixin.Events: [...] inherit from L.Evented instead.
- Remove the obsolete factory of Circle: https://leafletjs.com/reference.html#circle-l-circle

**Reply by johnd0e** [reply]
> * Remove the obsolete factory of Circle: https://leafletjs.com/reference.html#circle-l-circle

Not sure

**Comment by Malvoz** [comment]
`<button>` over `<a>` for controls: https://github.com/Leaflet/Leaflet/issues/7821

**Comment by www2000** [comment]
A late post from me but i like to see a arc function
see my feature request: #8005

**Reply by devgioele** [reply]
This sounds more appropriate as a plugin, isn't it?

**Reply by www2000** [reply]
Thanks for the response.
I see some canvas api functions like lineTo and Arc that replace X,Y cordinate with Lat and Lon is a good addition for Leaflet core api.

And with the current situation in the world be save.

**Comment by Malvoz** [comment]
Use the document's normal font size on `leaflet-container`: https://github.com/Leaflet/Leaflet/issues/8051

**Comment by dolorain** [comment]
support vector tilelayer

**Reply by iamtekson** [reply]
I found this [interesting research](https://www.mdpi.com/2220-9964/9/10/563). In this research, they compare the performance of three different popular open-source libraries, i.e. leafletjs, OpenLayers and maplibre. The result of this paper was quite surprising to me. The leaflet has the worst performance. Mainly they tried to load the vector tiles using the `vectorGrid` plugin. I think v2.0 should address this issue properly and load the vector tile layer in a faster way. Looking forward to seeing this feature in the v2.x.x. 

The summary of the research that I mentioned above is here: 

![zunino](https://github.com/Leaflet/Leaflet/assets/39838116/53ad5dcf-2c33-4533-abf5-8298945e4132)

**Comment by ansh-ag** [comment]
map rotation

**Reply by Malvoz** [reply]
Relevant info from https://github.com/Raruto/leaflet-rotate:

> _Initially based on the [work](https://github.com/Leaflet/Leaflet/tree/rotate) of Iván Sánchez Ortega (see: https://github.com/Leaflet/Leaflet/issues/268)_

**Reply by ansh-ag** [reply]
The plugin is okay, but its not as smooth as it should be. I dont remember correctly but there was a leaflet fork which had map rotation from few years, we used it to create a navigation(blue dot) experience. The only problem was if you do a panTo or panBy after rotation, the map will start to move in the wrong direction.

**Reply by imeidopf** [reply]
Google earth lets us set heading degrees. Rotate being implemented in the core, and better than the existing plugin, would be awesome to load direction-based content some way other than true north. E.g. a golf course has holes in all directions. I would rather look at the map according to my orientation to the existing whole.

**Comment by Danielku15** [comment]
**Support for Vector Tile Layers like MapBox GL including WebGL based rendering.** 

Looking at all the big Map providers (MapBox GL, Google Maps, Bing Maps etc.) this is the biggest eye catcher nowadays. Vector Tiles offer a lot in flexibility in the rendering of the map elements. Also it offers a much more dynamic display of elements like street labels (rotation, font size etc.). To provide such a professional and modern UX like the commercial players, this likely would imply a new WebGL based rendering backend. 

I'd love to see this within the leaflet core like raster tile layers. 

This would then also potentially be the foundation for future advanced features like map rotation, 3D Tilt, 3D object display (buildings) etc.  Potentially these things could/would come through plugins once the rendering core is there. 

I was very impressed with what MapBox GL provides, but it also comes with quite a bit of vendor lock. Seeing a bit more feature parity to it in Leaflet would make me 🤩

**Reply by mourner** [reply]
Not going to happen (saying this as someone who worked on Mapbox GL JS for 8 years, as well as creating Leaflet). Implementing good vector rendering engine is one of the most difficult things in engineering, requiring years of work by top engineers and millions in investments. For a nice modern library that renders vector tiles and works with Leaflet, I'd suggest checking out https://protomaps.com/

**Reply by IvanSanchez** [reply]
Vector tiles support is there - in the form of Leaflet.VectorGrid. Do not confuse "support for vector tiles" with "support for complex WebGL symbolization".

I've been doing stuff in the WebGL field since, what, 2015?, and can attest that it's stupidly hard. Only now I'm completing some work on https://gitlab.com/IvanSanchez/gleo , which *may* be adapted to be a Leaflet WebGL renderer. In the future. If things go according to plan. And, of course, no promises.

That's the main reason why I would like to see [this feature request of mine about "Extendable full-map-container DOM elements"](https://github.com/Leaflet/Leaflet/discussions/7729#discussioncomment-1578200) done for v3.

But a built-in complex webgl rendering engine as a core feature of Leaflet? Nope. That engine can **easily** take double the current amount of code Leaflet has.

**Reply by Danielku15** [reply]
Unfortunate to hear 😢. For me this is the future of web mapping where as raster tiles are rather a relic. We already moved to Canvas rendering due to the poor performance with DOM based rendering (we display quite some markers on our maps). As you pointed out nicely, it is a tough challenge which needs high level expertise in multiple domains. 

We might need to consider alternative libraries if Leaflet will not invest in this area. Doing a very basic competitor analysis shows that most other libraries bet on the horse of WebGL rendering (for 3D display and hardware acceleration purposes) and Vector Layers and Leaflet tends to lack behind there. I think almost all libraries nowadays have the basics like Raster Tiles, Custom Markers, WMS, Canvas 2D Rendering, GeoJSON, Custom Shapes, Attributions, Controls, Mobile support, extension APIs,... Of course every library with its strengths in some way (performance, lightweight, easy to use,..) but all providing a similar feature set.

But I also have to admit that we maybe have rather complex scenarios which might also go beyond the primary focus of leaflet being lightweight and easy to use. Those constraints might not be so much of relevance for us. 

My reference list: 
- Leaflet (Free, Open Source)
- Open Layers (Free, Open Source)
- MapBox GL JS (Free, Open Source, Affiliated with MapBox)
- Google Maps (Commercial, Closed Source, Affiliated with Google)
- Bing Maps (Commercial, Closed Source, Affiliated with Microsoft)
- Here Maps (Commercial, Closed Source, Affiliated with Here)
- Apple Maps (Commercial, Closed Source, Affiliated with Apple)

**Reply by SuperPat45** [reply]
I use leaflet with some plugins in combination with https://github.com/MapLibre/maplibre-gl-js (the fork of Mapbox GL JS v1.x) as a vector layer.
It works fine, except the JavaScript needed of the 2 API is very huge for small mobile connections.
Having leaflet v3 able to natively handle style.json vector layer will be awesome.
And it would be even better if he could rotate and tilt the map.

**Reply by iamtekson** [reply]
I found this [interesting research](https://www.mdpi.com/2220-9964/9/10/563). In this research, they compare the performance of three different popular open-source libraries, i.e. leafletjs, OpenLayers and maplibre. The result of this paper was quite surprising to me. The leaflet has the worst performance. Mainly they tried to load the vector tiles using the `vectorGrid` plugin. I have also tried to check the performance of vector tiles in different mapping APIs and found similar results. 

I think v2.0 should address this issue properly and load the vector tile layer in a faster way. Looking forward to seeing this feature in the v2.x.x. 

The summary of the research that I mentioned above is here: 

![zunino](https://github.com/Leaflet/Leaflet/assets/39838116/53ad5dcf-2c33-4533-abf5-8298945e4132)

**Comment by Malvoz** [comment]
SVGs for icons: https://github.com/Leaflet/Leaflet/issues/8158#issuecomment-1103727981 / https://github.com/Leaflet/Leaflet/issues/8756

**Comment by fox91** [comment]
Move from CSS to SCSS/SASS not with the intent of reducing the size of the generated CSS but to give plugins the ability to use the same style (colors, sizes, etc.) as the base components (#5832).

**Reply by millette** [reply]
Or simply <https://developer.mozilla.org/en-US/docs/Web/CSS/var> with <https://postcss.org/> (like babel for css).

**Comment by jonkoops** [comment]
I think it's time we created a branch for `v3` where we can accumulate breaking changes without influencing the day to day activity on `main`. This way we can keep landing fixes and other changes whilst having a longer lived branch for non-backwards compatible changes we want to get in.

WDYT @Malvoz @mourner @Falke-Design?

**Reply by Falke-Design** [reply]
I don't know if this would be good now. I think we have currently enough PRs and issues that should be resolved before we start working on `v3`. 
I personally think that `v3` can be earliest startet in an year and between that we have a few releases (> `1.9.0`) to get rid of the backlog

**Reply by mourner** [reply]
Well, a year is too far ahead — I hope we can move faster from now on. That said, we'd still need at least a few months for the current state to stabilize, patching up important issues, and it's super inconvenient to keep two running branches at once and cherry-pick every PR constantly. Currently there's no immediate need to introduce drastic changes.

**Reply by jonkoops** [reply]
Just FYI I can volunteer to keep the `v3` branch in-sync with `main`. I'd rebase it every other week or so. We should naturally try to keep any changes besides breaking ones in the main branch.

**Reply by Malvoz** [reply]
I'm commenting here because it is closely related:

I believe we should:

- rename this discussion to "Version 3 Features/Roadmap"
- create a new 3.0 milestone and probably move some things from [2.0](https://github.com/Leaflet/Leaflet/milestone/39), so that the 2.0 milestone can focus mainly on https://github.com/Leaflet/Leaflet/issues/6136.

thoughts @mourner, @IvanSanchez, @Falke-Design, @jonkoops?

Edit: this is also suggested in https://github.com/Leaflet/Leaflet/issues/6136#issuecomment-1115782867.

**Reply by Falke-Design** [reply]
> rename this discussion to "Version 3 Features/Roadmap"

Maybe we should not name it version 3. Who knows what else we need to release or if we switch generally to [semantic versioning ](https://semver.org/).

> so that the 2.0 milestone can focus mainly on #6136

Yes.
In my eyes the next release should be the 1.9 release with the things from [1.8.1 milestone](https://github.com/Leaflet/Leaflet/milestone/38) and maybe some other important patches. And then on the same day we Release 2.0 with dropping IE. After that step we also change our releases to [semantic versioning ](https://semver.org/). But this takes more planning because then we can't anymore merge what we want. We need to wait with breaking changes for a major release. Maybe someone (@IvanSanchez) has experience with that?

**Reply by Malvoz** [reply]
> > rename this discussion to "Version 3 Features/Roadmap"
> 
> Maybe we should not name it version 3. Who knows what else we need to release

But the features suggested here are supposedly (mostly) breaking changes (hence it was named v2?), so it'd have to be at least v3 right?

**Reply by Falke-Design** [reply]
> least v3 right?

Yes, but that are all big changes where we don't have a release date and I'm sure that we are not able to implement all features in the same release. Then we have always to rename this thread. A name which is not linked to a version would be much better. Like "Future of Leaflet (Features / Roadmap)"

**Comment by SuperPat45** [comment]
A globe view mode like the Google Maps one or Web GL Earth https://examples.webglearth.com/

**Comment by SuperPat45** [comment]
Add 'top-center', 'bottom-center' control positions of this pull request: https://github.com/Leaflet/Leaflet/pull/7930

**Reply by Malvoz** [reply]
This doesn't have to wait until v2 has been released.

**Comment by SuperPat45** [comment]
Use "window.devicePixelRatio" and remove L.Browser.Retina: https://github.com/Leaflet/Leaflet/issues/8188

**Comment by SuperPat45** [comment]
Prevent control to cover other controls: https://github.com/Leaflet/Leaflet/issues/7944 https://github.com/Leaflet/Leaflet/issues/4644
Maybe by using CSS grid layout?

**Reply by Malvoz** [reply]
- https://github.com/Leaflet/Leaflet/issues/8358.

**Comment by SuperPat45** [comment]
The trackResize option should use the ResizeObserver API instead of the window.on("reisze") event

**Comment by richardeschloss** [comment]
Hi, I really like the leaflet plugins but it seems like several depend on leaflet "L" being defined globally. For example, my favorite plugin https://github.com/MazeMap/Leaflet.TileLayer.PouchDBCached asks the user to add script tags before loading the plugin.

Would the team be open to supporting something like `L.use(plugin, options)` so that only a scoped instance of leaflet uses the plugins it defines? For example:

```js
import { L } from 'leaflet'
import TileLayerCache from 'leaflet.tileLayer.pouchDBCached'
L.use(TileLayerCache, { useCache: true })
```

**Reply by jonkoops** [reply]
I don't know if we want to implement such an API. We are getting rid of the globals by default in Leaflet 2 (with an opt-in to get them back), at that point we will consider it deprecated and plugins will need to move away from it.

**Reply by richardeschloss** [reply]
Great! I'm looking forward to Leaflet 2

**Comment by iamtekson** [comment]
Map `idle` event similar to Mapbox `idle` event.

**Reply by Malvoz** [reply]
- https://github.com/Leaflet/Leaflet/issues/3178
